sheets/api.py

```diff
@@ -310,6 +312,91 @@ def batch_update_sheet_cells(self, sheet_id, request_objects):
+            defined dynamically (ref: http://googleapis.github.io/google-api-python-client/docs/dyn/drive_v3.html)
+    """
+    credentials = credentials or get_credentials()
+    return build("drive", "v3", credentials=credentials, cache_discovery=False)
+
+
+def batch_share_callback(
+    request_id, response, exception
+):  # pylint: disable=unused-argument
+    """
+    A callback function given to the Google API client's new_batch_http_request(). Called for the result
+    of each individual request executed in a batch API call.
+    Ref: https://developers.google.com/drive/api/v3/batch#using-client-libraries
+    NOTE: If *any* of the individual share requests in the batch fail, this callback is called and passed an exception
+    for *all* of the individual requests whether they contained a bad email address or not. There is no way to
+    determine the specific failed emails from the arguments to this function.
+
+    Args:
+        request_id (str): The id of the individual batch request
+        response (dict): The API response body if no error occurred
+        exception (googleapiclient.errors.HttpError): The error, if one occurred
+    """
+    if exception:
+        raise FailedBatchRequestException from exception
+
+
+def share_drive_file_with_emails(file_id, emails_to_share, credentials=None):
+    """
+    Shares a Drive file with multiple recipients. First attempts to share with the list of emails via a batch request.
+    If that fails, attempts to share with each email address individually. If a batch share request fails, the file is
+    not shared with any of the emails, and there is no way to get reliable information about which specific emails
+    failed. This function defaults to individual sharing so valid emails will still be shared to if the batch request
+    fails.
+
+    Args:
+        file_id (str): The Drive file id
+        emails_to_share (list of str): Email addresses that will be added as shared users for the given file
+        credentials (google.oauth2.credentials.Credentials or None): Credentials to be used by the
+            Google Drive client
+
+    Returns:
+
+    """
+    drive_service = build_drive_service(credentials=credentials)
+    batch = drive_service.new_batch_http_request()
+    for email in emails_to_share:
+        batch.add(
+            drive_service.permissions().create(
+                **build_drive_file_email_share_request(file_id, email)
+            ),
+            request_id=email,
+            callback=batch_share_callback,
+        )
+    try:
+        batch.execute()
+    except FailedBatchRequestException:
+        log.exception(
+            "Failed to batch-share the spreadsheet '%s'. One or more of the emails failed, "
+            "so the spreadsheet was not shared with any of them. Now attempting to share individually..."
+        )
+        for email in settings.SHEETS_ADMIN_EMAILS:
```

Comment:
Based on the logging message here, shouldn't we be iterating over `emails_to_share`?